src/control/lib/hostlist/hostlist_test.go

```diff
@@ -0,0 +1,627 @@
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+package hostlist_test
+
+import (
+	"errors"
+	"reflect"
+	"testing"
+
+	"github.com/daos-stack/daos/src/control/lib/hostlist"
+)
+
+func makeStringRef(in string) *string {
+	return &in
+}
+
+func TestHostList_Create(t *testing.T) {
+	for name, tc := range map[string]struct {
+		startList    string
+		expRawOut    string
+		expUniqOut   string
+		expUniqCount int
+		expErr       error
+	}{
+		"simple": {
+			startList:    "node[1-128]",
+			expRawOut:    "node[1-128]",
+			expUniqOut:   "node[1-128]",
+			expUniqCount: 128,
+		},
+		"complex": {
+			startList:    "node2-1,node1-2,node1-[45,47],node3,node1-3",
+			expRawOut:    "node2-1,node1-[2,45,47],node3,node1-3",
+			expUniqOut:   "node3,node1-[2-3,45,47],node2-1",
+			expUniqCount: 6,
+		},
+		"complex with suffixes": {
+			startList:    "node2-1,node1-2.suffix1,node1-[45,47].suffix2,node3,node1-3",
+			expRawOut:    "node2-1,node1-2.suffix1,node1-[45,47].suffix2,node3,node1-3",
+			expUniqOut:   "node3,node1-3,node1-2.suffix1,node1-[45,47].suffix2,node2-1",
+			expUniqCount: 6,
+		},
+		"duplicates removed": {
+			startList:    "node[1-128],node2,node4,node8,node16,node32,node64,node128",
+			expRawOut:    "node[1-128,2,4,8,16,32,64,128]",
+			expUniqOut:   "node[1-128]",
+			expUniqCount: 128,
+		},
+		"compatible padding": {
+			startList:    "node[001-9],node[010-20]",
+			expRawOut:    "node[001-020]",
+			expUniqOut:   "node[001-020]",
+			expUniqCount: 20,
+		},
+		"incompatible padding": {
+			startList:    "node[001-9],node[10-20]",
+			expRawOut:    "node[001-009,10-20]",
+			expUniqOut:   "node[10-20,001-009]",
+			expUniqCount: 20,
+		},
+		"space delimited": {
+			startList:    "node1 node2 node3",
+			expRawOut:    "node[1-3]",
+			expUniqOut:   "node[1-3]",
+			expUniqCount: 3,
+		},
+		"tab delimited": {
+			startList: "node1	node2	node3",
+			expRawOut:    "node[1-3]",
+			expUniqOut:   "node[1-3]",
+			expUniqCount: 3,
+		},
+		"mixed delimiters": {
+			startList: "node1	node2 node3,node4",
+			expRawOut:    "node[1-4]",
+			expUniqOut:   "node[1-4]",
+			expUniqCount: 4,
+		},
+		"lo > hi": {
+			startList: "node[5-4]",
+			expErr:    errors.New("invalid range"),
+		},
+		"NaN-lo": {
+			startList: "node[a-b]",
+			expErr:    errors.New("invalid range"),
+		},
+		"NaN-hi": {
+			startList: "node[0-b]",
+			expErr:    errors.New("invalid range"),
+		},
+		"bad range": {
+			startList: "node[0--1]",
+			expErr:    errors.New("invalid range"),
+		},
+		"weird": {
+			startList: "node[ab]",
+			expErr:    errors.New("invalid range"),
+		},
+		"unclosed range": { // seems like it should be an error, but C version allows it
+			startList:    "node[0-1",
+			expRawOut:    "node[0-1",
+			expUniqOut:   "node[0-1",
```

Comment:
Would it make sense to correct this for our structure's output? Even if the C version accepts it, it's pretty ugly.

Reply:
My thinking (admittedly while on cold medicine) was to be conservative about deviating from the C implementation, but on further reflection I agree with this. It doesn't seem useful to allow it.